apache/superset
# feat: add permalink to dashboard and explore

### SUMMARY

This PR adds new permanent link functionality that makes it possible to persist Dashboard and Explore state in the metadata database. The state is persisted in a new metadata table called `key_value`, which is able to store pickleable binary objects (can be used in the future for other resources, too). The new permalinks look as follows:
- Dashboard: `http://host/superset/dashboard/p/<key>/`
- Explore: `http://host/superset/explore/p/<key>/`

The dashboard state is a `dict` with the following fields:
- `filterState`: native filter state
- `hash` (optional): the anchor (used for direct links to charts/tabs)
- `urlParams` (optional): any URL params that are active, excluding the reserved URL params like `native_filter_key` etc. This is also where filter box state is persisted. Note that these are stored as a list of key-value tuples to support having multiple same keys (not currently supported by the chart data API but implemented like this for future proofing)

The Explore state is also a `dict` with the following fields:
- `formData` the chart form data
- `urlParams` (optional): any URL params that are active, excluding the reserved URL params like `form_data_key`, `slice_id` etc. (same format as in the dashboard state)

The benefits compared to the deprecated `shorturl` functionality:
- Leverages existing resource security mechanism: if you don't have access to a chart/dashboard, you are also not able to view permalink state for those charts/dashboards
- Uses UUID-based keys by default to make permalinks unguessable. However, where needed, numeric serial keys (similar to the legacy shorturl keys) can be used by flipping a config flag. Do note, however, that changing this setting will cause old links to stop working, i.e. it's not possible to retrieve permalink state based by its UUID key if the feature has been configured to work with the serial id and vice versa.
- Is not restricted by the URL length limitations that affects shorturls (however, Filter Box state is still subject to URL overflows).

Notice that old shorturls will still work like before - but links created after this will be based on the new permalink feature.

Detailed list of changes:
- Rename current `key_value` package to `temporary_cache` to better reflect the ephemeral nature of the state (not guaranteed to be persisted)
- Add new `key_value` table to metadata database along with relevant commands for adding, updating, getting and deleting key-value pairs
- Add `permalink` endpoints (POST and GET) to both Explore and Dashboard that leverage the new `key_value` store
- Add `PERMALINK_KEY_TYPE` to `config.py` which makes it possible to switch between serial and UUID based keys (UUID = default)
- fully support persisting filter box state and url parameters in permalink state (this was currently mostly broken)
- remove old shorturl creation functionality from frontend and backend
- Migrate all permalink buttons to new feature:
  - Explore action buttons (above chart): "Copy permalink to clipboard" and "Share permalink by email" 
  - Explore embed iframe button (above chart): link now references a permalink
  - Dashboard context menu: "Copy permalink to clipboard" and "Share permalink by email"
  - Dashboard chart context menu: "Copy permalink to clipboard" and "Share permalink by email" 
  - Dashboard tab anchor link

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/157868216-4ce210b1-137b-4fa0-b9b3-39ddf67bc98d.mov

### TESTING INSTRUCTIONS
1. Explore a dataset, add a metric/columns to a few controls and share a permalink both via clipboard and email. Change control values and verify that changes don't change the state in the previously shared permalink. Bonus: create a chart with Jinja that uses `url_param` and make sure URL params are correctly persisted in permalinks.
2. Click the Embed code button and make sure the link works
3. Save the chart and ensure sharing works similar to above
4. Go to a dashboard with native filters and tabs, select a few filters and share the dashboard from the context menu.
5. Click on the anchor button next to a tab and make sure both the link and email message works
6. Click on a chart context menu and make sure both link and email message works
7. Create a dashboard with a filter box and a chart, apply a filter on the filter box and make sure shared permalinks open with the predefined filters
8. Bonus: add a chart with Jinja that references a `url_param` and make sure the chart renders correctly 1) in the dashboard after creating a permalink 2) in the explore view after creating a permalink to the chart.

### ADDITIONAL INFORMATION
